daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/TransactionPreprocessor.scala

```diff
@@ -95,28 +61,52 @@ private[preprocessing] final class TransactionPreprocessor(
       tx: GenTransaction[NodeId, Cid]
   ): (ImmArray[speedy.Command], Set[ContractId]) = {
 
-    type Acc = ((Set[Value.ContractId], Set[Value.ContractId]), BackStack[speedy.Command])
+    type Acc = (Set[Value.ContractId], Set[Value.ContractId], BackStack[speedy.Command])
 
-    val ((localCids, _), cmds) =
-      tx.roots.foldLeft[Acc](((Set.empty, Set.empty), BackStack.empty)) {
-        case ((cids, stack), id) =>
-          tx.nodes.get(id) match {
+    val (_, globalCids, cmds) =
+      tx.roots.foldLeft[Acc]((Set.empty, Set.empty, BackStack.empty)) {
+        case ((localCids0, globalCids0, stack), id) =>
+          val (newGlobals, newLocals, cmd) = tx.nodes.get(id) match {
             case None =>
               fail(s"invalid transaction, root refers to non-existing node $id")
             case Some(node) =>
               node match {
+                case create: Node.NodeCreate[Cid] =>
+                  val (cmd, newCids) =
+                    commandPreprocessor.unsafePreprocessCreate(create.templateId, create.arg)
+                  val newGlobalCids = newCids.filterNot(localCids0)
+                  (newGlobalCids, List(create.coid), cmd)
+                case exe: Node.NodeExercises[_, Cid] =>
+                  val templateId = exe.templateId
+                  val (cmd, newCids) =
+                    commandPreprocessor.unsafePreprocessExercise(
+                      templateId,
+                      exe.targetCoid,
+                      exe.choiceId,
+                      exe.chosenValue,
+                    )
+                  val newGlobalCids = newCids.filterNot(localCids0)
+                  val newLocalCids = GenTransaction(tx.nodes, ImmArray(id)).localContracts.keys
+                  (newGlobalCids, newLocalCids, cmd)
+
                 case _: Node.NodeFetch[_] =>
                   fail(s"Transaction contains a fetch root node $id")
                 case _: Node.NodeLookupByKey[_] =>
                   fail(s"Transaction contains a lookup by key root node $id")
-                case _ =>
-                  val (cmd, acc) = unsafeTranslateNode(cids, node)
-                  (acc, stack :+ cmd)
+                case _: Node.NodeRollback[_] =>
+                  // TODO https://github.com/digital-asset/daml/issues/8020
+                  // how on earth can we turn a rollback node back into a speedy command?
+                  sys.error("rollback nodes are not supported")
               }
           }
+          val globalCids = globalCids0 | newGlobals
```

Comment:
Are you sure that this works? I'm a bit uneasy that we can't distinguish the following two scenarios:

Scenario 1:
1. There is a contract with cid `cid0` on the ledger.
2. We submit a two-command transaction: Command 1's transaction tree contains a create node that creates a contract with local cid `cid0`. Command 2 is a plain exercise on the contract `cid0` that exists on the ledger. However, replay now interprets this command as acting on the newly created contract.

Scenario 2:
1. There is no contract with cid `cid0` on the ledger.
2. We submit a two-command transaction: Command 1's transaction tree contains a create node that creates a contract with local cid `cid0` and key `k`. Command 2 is a by-key exercise on the key `k`.

One could probably argue that scenario 1 should have failed during interpretation in the submission service, but I'm worried that this subtlety gets overlooked at some point in the future and we get the next bug here.

Reply:
That is a good point, we definitely violate the spec here, quoting:
```
* The *local* contract IDs are the IDs of the contracts created by the
  transaction.

* The *global* contract IDs are the contract IDs that:
   
   * appear in the commands that produced the transaction. This
     includes the IDs of the exercised contract, together with all the
     IDs referenced in the arguments of the create and exercise
     commands;
   * that are fetched or looked up by key unless they are local;
   * are referenced in the input contracts.
```
so `cid0` should definitely be global due to the exercise and we should get a conflict.

Not really sure how to fix that. It seems like the only way out is to preserve `byKey` info on exercise which we should do either way but that doesn’t help for old transaction versions.

Reply:
After another offline discussion with @remyhaemmerle-da: There are probably two ways to approach this;

1. Properly distinguish between normal exercise and exerciseByKey nodes using the `byKey` flag. For backwards compatibility, we could treat all exercise nodes without a `byKey` field as normal exercises. This preserves the current behaviour, but still needs special treatment for `CreateAndExercise` commands.
2. In general allow exercises in later commands to use contract IDs created in earlier commands of the same transaction (both as input contract and in their choice arguments). This is the direction of this PR.

In scenario 1, the created contract clashes with an existing contract. So the Daml ledger should reject the submission with `Contract cid0 already exists`. The interesting part here is that the existing `cid0` might have been archived long ago (and just sits in the submitting participant's store of divulged contracts, so pruning may have removed the existence of `cid0` from the ledger's memory, but then the skew bounds on the ledger time (a contract ID commits to the ledger time of the creating transaction) should be tight enough that `cid0` cannot be created in this transaction.

So, we could update the spec to say that if a later exercise command refers to a contract created by an earlier command, then that's not considered a global contract ID for the transaction. What should we do about an exercise command referring to a contract ID that's created in its consequences or by a later command? I'd argue that this is an internally inconsistent transaction and should be rejected, and AFAICS this PR implements this.

Reply:
I think there are two mismatches which bite us here both of which we should fix but they are slightly orthogonal:
1. The mismatch between transaction nodes and LF primitives. This affects by-key flags but not createAndExercise (there is no primitive for that in LF)
2. The mismatch between transaction nodes and ledger API commands. This affects by-key flags but only on root nodes and createAndExercise.

We should absolutely fix 1 and add by-key flags but as you pointed out that does not make the issue go away for createAndExercise. Just doing that leads to what I’d argue are slightly unintuitive semantics: If you modify your scenario 1 such that the exercise in the second command is not the root node, this is already perfectly fine. Your approach 2 (implemented in this PR but probably should also get a spec update) removes that annoying special treatment of root nodes from the cid freshness check afaict. That seems like a more robust solution to address mismatch 2 (in this instance). I’m not really qualified to judge if relaxing the CID check here causes any issues but it sounds like you’re saying it’s fine. Assuming that is the case, my preference would be the following two steps:

1. Implement approach 2, i.e., this PR but also update the spec to match this (can be in a separate PR). This is all we need to fix the CID freshness check.
2. Add byKey flags to transaction nodes after that (soonish). This address mismatch 1 which causes other issues.

wdyt?

Reply:
I don't see a problem with relaxing the CID check here, as this should be caught on a higher level (freshness check for created contracts). I recommend to document the need for this freshness check; otherwise we'll forget and eventually run into a problem with some of the Daml ledgers.

The `byKey` flag is orthogonal in this approach. I would like to see the flag added officially at some point; it's not just at root nodes though, because child nodes can be come root nodes via projection.